[dvsim] Initial verible lint integration

msfschaffner opened the pull request:
This allows us to call Verible through dvsim and produce a dashboard similar to the one that we produce for Ascentlint. The report parsing script has been largely reused from the Ascentlint flow.
See [here](https://reports.opentitan.org/hw/top_earlgrey/lint/veriblelint/summary.html) for a sneak preview of the dashboard. I am going to add this to my lint cronjob once this PR is merged.

Note that this flow still produces many warnings, since we do not yet make use of the new configurability features that have been added to Verible (#229). For that we have to make sure we can pass the configuration file to Verible through Fusesoc by extending the Edalize backend:
https://github.com/olofk/edalize/blob/63d9512f29dc552099492f10cc45ae0df1490a9f/edalize/veriblelint.py#L64

The "run-in-bash" wrapper is currently needed in order to write the tool output to a logfile. If there would be another way to write such an output file that would of course be great. We mainly have two options to achieve this: we could either add that functionality to Verible or the Fusesoc command wrapper (I tend to prefer the former if that is possible). 

For reference, Ascentlint generates two output files that are of interest to us: 

1) a `.log` file with general tool errors/warnings, including messages from syntax checking.
2) a `.rpt` file with lint-rule-specific messages with different severities (info, warning, error)

Example of Ascentlint invocation in the Edalize backend:
https://github.com/olofk/edalize/blob/63d9512f29dc552099492f10cc45ae0df1490a9f/edalize/templates/ascentlint/Makefile.j2#L7-L8

CC @hzeller @mithro @fangism 

Signed-off-by: Michael Schaffner <msf@opentitan.org>

sriyerg commented on hw/top_earlgrey/lint/ascentlint/top_earlgrey_lint_cfgs.hjson:
```diff
@@ -14,127 +14,127 @@
                   fusesoc_core: lowrisc:ip:aes
                   import_cfgs: ["{proj_root}/hw/lint/data/common_lint_cfg.hjson"]
-                  rel_path: "hw/ip/aes/lint"
+                  rel_path: "hw/ip/aes/lint/ascentlint"
```
Move this file to `hw/top_earlgrey/lint` and change this to `rel_path: "hw/ip/aes/lint/{tool}`. That way you only need one copy. Set `rel_path` for the summary accordingly as well.

msfschaffner replied:
ok, makes sense, thanks.

sjgitty commented:
> See here for a sneak preview of the dashboard. 

Can you make sure this is listed as either `verible lint` or `style lint`
to distinguish from the RTL lint?

sriyerg commented on hw/lint/data/common_lint_cfg.hjson:
```diff
@@ -17,11 +13,11 @@
 
   // we rely on fusesoc to run lint for us
-  build_cmd:  "fusesoc "
+  build_cmd:  "{tool_srcs_dir}/run-in-bash.sh {build_dir}/lint-{tool}/{tool}-tool.log fusesoc "
```
How about we add `build_log` variable in the Hjson and supply it to `lint.mk`? We could change [lint.mk](https://github.com/lowRISC/opentitan/blob/6b869134f355fe3e6e79d64a7a5e4b33480e3e7b/hw/lint/data/lint.mk#L22) to this:
```
	-cd ${build_dir} && ${build_cmd} ${build_opts} | tee ${build_log}
```

msfschaffner replied:
good idea. I'll do that for synthesis as well (#1749), since DC does not support writing logfiles for the main shell except for a command log, which is of no use in this case.

msfschaffner replied:
Ok, I have added a mechanism to override the default title postfix which is "Lint Results" for the lint class. For Verible, I now override this with "Style Lint Results". Also, if you click on one of the modules on the dashboard, the page that comes up with the detailed messages also indicates which tool was used (this would be "veriblelint" in this case).

@sriyerg I have added a field named `results_title_postfix` to the `OneShotCfg` class. If that field is empty, the default string of each subclass is used, e.g. "Lint Results"  or "Synthesis Results". Let me know if you think we should do this differently. We can also add this to the base class if this could be useful for other flows.

imphil commented on hw/lint/tools/veriblelint/parse-lint-report.py:
```diff
@@ -0,0 +1,115 @@
+from pathlib import Path
+
+import hjson
+import mistletoe
```
that's unused

msfschaffner replied:
removed, thanks.

imphil commented on hw/lint/tools/veriblelint/parse-lint-report.py:
```diff
@@ -0,0 +1,115 @@
+r"""Parses lint report and dump filtered messages in hjson format.
+"""
+import argparse
+import datetime
```
unused

msfschaffner replied:
dito

msfschaffner commented:
Thanks for the comments! I'll update this PR once #1749 has made it in, since I will need features that have been added there.

msfschaffner commented:
> * Please file an issue with Verible to get this functionality added, add big TODOs with the upstream issue number to the current code, and file a follow-up issue here in OT to get that workaround removed once upstream adds this functionality. Ideally, they directly add structured reporting as well to save us from fragile text file parsing.
> * If we'd want to keep that longer than a couple weeks, I'd at least change the `sim.mk` file to get this functionality in there, and avoid the need for a wrapper script.

Thanks for giving this a look. I had a discussion about this with @sriyerg, and it appears that we need some sort of `tee` functionality, since we found that also other tools do not have this functionality built in. For instance, while Synopsys DC can write a command log and all sorts of report files, it looks like there is no way to write a complete DC shell log.

Hence, we decided to add this functionality to the corresponding `{flow}.mk` file, and this allows to get rid of this ugly workaround I have used here. The functionality is already in place as I have added that in one of the synthesis-related PRs last week (#1749). 

I have heard that we may eventually get some sort of structured result output from Verible, which will make the result parsing script superfluous. I'll check again with the authors (also regarding common log files) in the next meeting that we have with them. In the meantime, I have added a tracking item to https://github.com/lowRISC/opentitan/issues/1727 for structured tool output support.

imphil commented on hw/lint/data/ascentlint.hjson:
```diff
@@ -0,0 +1,17 @@
+                 "run --target={flow} --tool={tool} --build-root={build_dir}",
+                 " {fusesoc_core} "]
+
+    // Ascentlint-specific results parsing script that is called after running synthesis
```
```suggestion
    // Ascentlint-specific results parsing script that is called after running lint
```

imphil commented on hw/lint/data/veriblelint.hjson:
```diff
@@ -0,0 +1,21 @@
+    build_cmd:  "fusesoc "
+    build_opts: ["--cores-root {proj_root}",
+                 "run --target={flow} --tool={tool} --build-root={build_dir}",
+                 " {fusesoc_core} "]
```
as above: please split at space into individual arguments.

msfschaffner replied:
dito

imphil commented on hw/lint/data/veriblelint.hjson:
```diff
@@ -0,0 +1,21 @@
+                 "run --target={flow} --tool={tool} --build-root={build_dir}",
+                 " {fusesoc_core} "]
+
+    // TODO(#1727): switch over to native structured tool output, once supported by Verible
```
Let's please not link a catch-all issue here, that's impossible to track and action. Ideally, we can link a Verible issue?

msfschaffner replied:
I've added it to the DVSIM tracker since the output parsing script is DVSIM specific, but I can also add a point here in this Verible-related issue: https://github.com/lowRISC/opentitan/issues/1342

msfschaffner replied:
I've carried these TODOs over to the Verible Issue #1342

imphil commented on hw/lint/data/veriblelint.hjson:
```diff
@@ -0,0 +1,21 @@
+                 " {fusesoc_core} "]
+
+    // TODO(#1727): switch over to native structured tool output, once supported by Verible
+    // Verible-specific results parsing script that is called after running synthesis
```
```suggestion
    // Verible lint-specific results parsing script that is called after running lint
```

imphil commented on hw/lint/tools/veriblelint/parse-lint-report.py:
```diff
@@ -0,0 +1,111 @@
+    parser.add_argument('--outdir',
+                        type=str,
+                        default="./",
+                        help="""Output directory for the 'results.hjson' file.
+                        Defaults to './'""")
```
```suggestion
                        help="Output directory for the 'results.hjson' file. Defaults to '%(default)s'")
```

msfschaffner commented on hw/lint/data/ascentlint.hjson:
```diff
@@ -0,0 +1,17 @@
+    tool_srcs: []
+
+    // We rely on fusesoc to run lint for us
+    build_cmd:  "fusesoc "
```
let me double check that. I have taken this from already existing scripts. I hope that this is not the case neither :), otherwise I'll fix that.

msfschaffner replied:
fixed (also other occurrences in lint and synthesis targets).

msfschaffner commented on hw/lint/data/ascentlint.hjson:
```diff
@@ -0,0 +1,17 @@
+    build_cmd:  "fusesoc "
+    build_opts: ["--cores-root {proj_root}",
+                 "run --target={flow} --tool={tool} --build-root={build_dir}",
+                 " {fusesoc_core} "]
```
sure, will do.

msfschaffner commented:
Thanks @imphil for the review! Waiting for @sriyerg's final approval before merging in.

sriyerg commented on hw/lint/data/common_lint_cfg.hjson:
```diff
@@ -6,22 +6,14 @@
-  build_log: "{build_dir}/{dut}_flow.log"
-
-  report_cmd: "{proj_root}/hw/lint/tools/{tool}/parse-lint-report.py "
-  report_opts: ["--repdir={build_dir}/lint-{tool} --outdir={build_dir}"]
-  tool_srcs: []
+  build_log: "{build_dir}/lint.log"
```
nit: please align and group like items together (build_dir and build_log).

msfschaffner replied:
`build_dir` and `build_log`  are already grouped together, but the alignment of the strings on the right hand side is not correct. will fix.

sriyerg commented on hw/lint/data/veriblelint.hjson:
```diff
@@ -0,0 +1,24 @@
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+{
+    tool_srcs: []
+
+    // TODO(#1342): pass Verible config file to FuseSoC, once supported by Verible
+    // We rely on fusesoc to run lint for us
+    build_cmd:  "fusesoc"
+    build_opts: ["--cores-root {proj_root}",
+                 "run",
+                 "--target={flow}",
+                 "--tool={tool}",
+                 "--build-root={build_dir}",
+                 "{fusesoc_core}"]
+
+    // TODO(#1342): switch over to native structured tool output, once supported by Verible
+    // Verible lint-specific results parsing script that is called after running lint
+    report_cmd: "{proj_root}/hw/lint/tools/{tool}/parse-lint-report.py "
+    report_opts: ["--repdir={build_dir}",
+                  "--outdir={build_dir}"]
```
This seems exactly identical to ascentlint cfg.

msfschaffner replied:
This actually was located in the common lint config file initially, but I have split this out since the report command options are different for Verible, and we plan to update both the report command and the build command in the near future (mind the TODO's ;)). 

But thinking about this some more, I think we can leave the FuseSoC build command aligned, since the Verible configuration file will likely be defined in a common FuseSoC lint core file (similar to how this has been done for AscentLint). 

Will adjust that.

sriyerg replied:
Okay, sounds good.

sriyerg commented on hw/top_earlgrey/lint/top_earlgrey_lint_cfgs.hjson:
```diff
@@ -6,135 +6,140 @@
                   fusesoc_core: lowrisc:ip:aes
                   import_cfgs: ["{proj_root}/hw/lint/data/common_lint_cfg.hjson"]
-                  rel_path: "hw/ip/aes/lint"
+                  rel_path: "hw/ip/aes/lint/{tool}"
```
Thanks, this is exactly what I had in mind.

sriyerg commented on hw/lint/data/veriblelint.hjson:
```diff
@@ -0,0 +1,24 @@
+    report_opts: ["--repdir={build_dir}",
+                  "--outdir={build_dir}"]
+
+    results_title_postfix: " Style Lint Results"
```
Is it better to convert this to a flag-like structure:
```
  is_style_lint: True
```
Based on this value, you can update the behavior in the `LintCfg` class (at more than one place even, if needed).

msfschaffner replied:
I was thinking that a postfix field would give us some more flexibility for future tools (we may want to add CDC or RDC lint passes in the future). But if you think we should just enumerate all these lint types I'll be happy to change that.

msfschaffner replied:
I think you are right, especially if we want to customize the reports beyond just changing the title in the future. I've converted this to a flag.

sriyerg replied:
In any case, I am fine with either solution.

msfschaffner commented:
Thanks @sriyerg, updated.
